File: tests/oracle/type_unknown_1204.cc
/*
 * TYPE-UNKNOWN-0001: locked Ghidra 12.0.4 TypeFactory unknown-base oracle.
 */
#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "type.hh"

#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;

void writeString(ostream &out,const string &value)
{
  out << '"';
  for(size_t i=0;i<value.size();++i) {
    unsigned char ch = static_cast<unsigned char>(value[i]);
    if (ch == '"' || ch == '\\') out << '\\' << static_cast<char>(ch);
    else if (ch == '\n') out << "\\n";
    else out << static_cast<char>(ch);
  }
  out << '"';
}

void writeType(ostream &out,int4 requestedSize,Datatype *type)
{
  string metatype;
  metatype2string(type->getMetatype(),metatype);
  out << "{\"requested_size\":" << requestedSize
      << ",\"size\":" << type->getSize()
      << ",\"metatype\":";
  writeString(out,metatype);
  out << ",\"name\":";
  writeString(out,type->getName());
  out << ",\"id\":" << type->getId()
      << ",\"flags\":" << type->getInheritable()
      << '}';
}

void writeAnonymousUnknownOrder(ostream &out,TypeFactory *factory)
{
  vector<Datatype *> ordered;
  factory->dependentOrder(ordered);
  bool first = true;
  out << '[';
  for(size_t i=0;i<ordered.size();++i) {
    Datatype *type = ordered[i];
    if (type->getMetatype() != TYPE_UNKNOWN || !type->getName().empty())
      continue;
    if (!first) out << ',';
    out << type->getSize();
    first = false;
  }
  out << ']';
}

void runFixture(const string &specDirectory,const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  BfdArchitecture architecture(binary,"default",&std::cerr);
  DocumentStorage store;
  architecture.init(store);

  TypeFactory *factory = architecture.types;
  if (factory == (TypeFactory *)0)
    throw std::runtime_error("architecture has no TypeFactory");

  const int4 sizes[] = { 8, 1, 4, 2, 3, 5, 6, 7 };
  vector<Datatype *> first;
  std::cout << "{\"schema\":1,\"fixture\":\"TYPE-UNKNOWN-0001\",\"types\":[";
  for(size_t i=0;i<sizeof(sizes)/sizeof(sizes[0]);++i) {
    if (i != 0) std::cout << ',';
    Datatype *type = factory->getBase(sizes[i],TYPE_UNKNOWN);
    first.push_back(type);
    writeType(std::cout,sizes[i],type);
  }
  std::cout << "],\"identity\":{\"repeat\":[";
  for(size_t i=0;i<first.size();++i) {
    if (i != 0) std::cout << ',';
    std::cout << (first[i] == factory->getBase(sizes[i],TYPE_UNKNOWN));
  }
  std::cout << "],\"different_size\":[";
  for(size_t i=1;i<first.size();++i) {
    if (i != 1) std::cout << ',';
    std::cout << (first[0] == first[i]);
  }
  std::cout << "]},\"named\":";

  Datatype *named = factory->getBase(3,TYPE_UNKNOWN,"fixture_unknown3");
  writeType(std::cout,3,named);
  std::cout << ",\"named_repeat\":"
            << (named == factory->getBase(3,TYPE_UNKNOWN,"fixture_unknown3"));
  std::cout << ",\"collision_error\":";
  try {
    (void)factory->getBase(4,TYPE_UNKNOWN,"fixture_unknown3");
    writeString(std::cout,"NONE");
  }
  catch(const LowlevelError &err) {
    writeString(std::cout,err.explain);
  }

  std::cout << ",\"anonymous_order_before_clear\":";
  writeAnonymousUnknownOrder(std::cout,factory);

  factory->clearNoncore();
  std::cout << ",\"anonymous_order_after_clear\":";
  writeAnonymousUnknownOrder(std::cout,factory);
  Datatype *afterClear = factory->getBase(3,TYPE_UNKNOWN);
  std::cout << ",\"clear_identity\":{\"new_repeat\":"
            << (afterClear == factory->getBase(3,TYPE_UNKNOWN))
            << "},\"anonymous_order_after_recreate\":";
  writeAnonymousUnknownOrder(std::cout,factory);
  std::cout << "}\n";
}

} // namespace

int main(int argc,char **argv)
{
  try {
    if (argc != 3)
      throw std::invalid_argument("usage: type_unknown_1204 SPEC_DIRECTORY BINARY");
    runFixture(argv[1],argv[2]);
    return 0;
  }
  catch(const ghidra::LowlevelError &err) {
    std::cerr << "LowlevelError: " << err.explain << '\n';
  }
  catch(const std::exception &err) {
    std::cerr << "std::exception: " << err.what() << '\n';
  }
  return 1;
}

File: tests/oracle/type_unknown_1204.rs
use std::sync::Arc;

use rugra::type_system::datatype::{type_flags, Datatype, TypeMetatype};
use rugra::type_system::typefactory::TypeFactory;

fn write_json_string(out: &mut String, value: &str) {
    out.push('"');
    for ch in value.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            _ => out.push(ch),
        }
    }
    out.push('"');
}

fn metatype_name(metatype: TypeMetatype) -> &'static str {
    match metatype {
        TypeMetatype::Unknown => "unknown",
        _ => panic!("fixture received a non-unknown type"),
    }
}

fn write_type(out: &mut String, requested_size: usize, datatype: &Datatype) {
    out.push_str("{\"requested_size\":");
    out.push_str(&requested_size.to_string());
    out.push_str(",\"size\":");
    out.push_str(&datatype.get_size().to_string());
    out.push_str(",\"metatype\":");
    write_json_string(out, metatype_name(datatype.get_metatype()));
    out.push_str(",\"name\":");
    write_json_string(out, datatype.get_name());
    out.push_str(",\"id\":");
    out.push_str(&datatype.get_id().to_string());
    out.push_str(",\"flags\":");
    out.push_str(&(datatype.get_flags() & type_flags::CORETYPE).to_string());
    out.push('}');
}

fn write_anonymous_unknown_order(out: &mut String, factory: &TypeFactory) {
    out.push('[');
    let mut first = true;
    let mut ordered = Vec::new();
    factory.dependent_order(&mut ordered);
    for datatype in ordered {
        if datatype.get_metatype() != TypeMetatype::Unknown || !datatype.get_name().is_empty() {
            continue;
        }
        if !first {
            out.push(',');
        }
        out.push_str(&datatype.get_size().to_string());
        first = false;
    }
    out.push(']');
}

fn main() {
    // The locked oracle drives SleighArchitecture's standalone
    // buildCoreTypes (sleigh_arch.cc:229-232), so the Rust comparand
    // constructs the factory with the same registration flavor.
    let mut factory = TypeFactory::new_flavor(8, rugra::type_system::typefactory::CoreTypeFlavor::Standalone);
    let sizes = [8_usize, 1, 4, 2, 3, 5, 6, 7];
    let first: Vec<Arc<Datatype>> = sizes
        .iter()
        .map(|&size| {
            factory
                .get_base(size, TypeMetatype::Unknown)
                .expect("unknown base type must exist")
        })
        .collect();

    let mut out = String::from(
        "{\"schema\":1,\"fixture\":\"TYPE-UNKNOWN-0001\",\"types\":[",
    );
    for (index, (&size, datatype)) in sizes.iter().zip(first.iter()).enumerate() {
        if index != 0 {
            out.push(',');
        }
        write_type(&mut out, size, datatype);
    }

    out.push_str("],\"identity\":{\"repeat\":[");
    for (index, (&size, datatype)) in sizes.iter().zip(first.iter()).enumerate() {
        if index != 0 {
            out.push(',');
        }
        let repeated = factory
            .get_base(size, TypeMetatype::Unknown)
            .expect("repeated unknown base type must exist");
        out.push(if Arc::ptr_eq(datatype, &repeated) { '1' } else { '0' });
    }

    out.push_str("],\"different_size\":[");
    for (index, datatype) in first.iter().enumerate().skip(1) {
        if index != 1 {
            out.push(',');
        }
        out.push(if Arc::ptr_eq(&first[0], datatype) { '1' } else { '0' });
    }
    out.push_str("]},\"named\":");

    let named = factory
        .get_base_named(3, TypeMetatype::Unknown, "fixture_unknown3")
        .expect("named unknown type must be created");
    write_type(&mut out, 3, &named);
    let repeated_named = factory
        .get_base_named(3, TypeMetatype::Unknown, "fixture_unknown3")
        .expect("named unknown type must be canonical");
    out.push_str(",\"named_repeat\":");
    out.push(if Arc::ptr_eq(&named, &repeated_named) {
        '1'
    } else {
        '0'
    });
    out.push_str(",\"collision_error\":");
    let collision = factory
        .get_base_named(4, TypeMetatype::Unknown, "fixture_unknown3")
        .map(|_| "NONE".to_string())
        .unwrap_or_else(|error| error);
    write_json_string(&mut out, &collision);

    out.push_str(",\"anonymous_order_before_clear\":");
    write_anonymous_unknown_order(&mut out, &factory);

    factory.clear_non_core();
    out.push_str(",\"anonymous_order_after_clear\":");
    write_anonymous_unknown_order(&mut out, &factory);
    let after_clear = factory
        .get_base(3, TypeMetatype::Unknown)
        .expect("unknown base must be recreated after clear");
    out.push_str(",\"clear_identity\":{\"new_repeat\":");
    let after_clear_repeat = factory
        .get_base(3, TypeMetatype::Unknown)
        .expect("recreated unknown base must be canonical");
    out.push(if Arc::ptr_eq(&after_clear, &after_clear_repeat) {
        '1'
    } else {
        '0'
    });
    out.push_str("},\"anonymous_order_after_recreate\":");
    write_anonymous_unknown_order(&mut out, &factory);
    out.push_str("}\n");
    print!("{out}");
}
